app/scripts/metamask-controller.js

```diff
@@ -3494,41 +3494,29 @@ export default class MetamaskController extends EventEmitter {
         this._convertMnemonicToWordlistIndices(seedPhraseAsBuffer),
       );
 
-      // Scan accounts until we find an empty one
-      const { chainId } = this.networkController.state.providerConfig;
+      // Scan accounts and apply checks
       const ethQuery = new EthQuery(this.provider);
       const accounts = await this.keyringController.getAccounts();
-      let address = accounts[accounts.length - 1];
-
-      for (let count = accounts.length; ; count++) {
-        const balance = await this.getBalance(address, ethQuery);
+      let emptyAccountsCounter = 0;
 
-        if (balance === '0x0') {
-          // This account has no balance, so check for tokens
-          await this.detectTokensController.detectNewTokens({
-            selectedAddress: address,
-          });
-
-          const tokens =
-            this.tokensController.state.allTokens?.[chainId]?.[address];
-          const detectedTokens =
-            this.tokensController.state.allDetectedTokens?.[chainId]?.[address];
+      for (
+        let index = accounts.length - 1;
+        emptyAccountsCounter < 10;
+        index++
+      ) {
+        let address = accounts[index];
+        const isAccountEmpty = await this.isAccountEmpty(address, ethQuery);
 
-          if (
-            (tokens?.length ?? 0) === 0 &&
-            (detectedTokens?.length ?? 0) === 0
-          ) {
-            // This account has no balance or tokens
-            if (count !== 1) {
-              await this.removeAccount(address);
-            }
-            break;
+        if (isAccountEmpty) {
+          emptyAccountsCounter += 1;
+          if (index !== 0) {
+            await this.removeAccount(address);
```

Comment:
I don't think accounts can be removed while we're iterating?  If I remove account `<N>` and try to add `<N+1>`, it throws "Account out of sequence...".  I think it may have to remove them all at the end, after the loop?

Reply:
I have not looked deeply into this yet, but I suspect that if we want to do what this code intends on doing, we need to make changes to the keyring controller. For example, I don't think we can remove an arbitrary set of accounts, we can only remove accounts sequentially... although maybe recent changes to the controller (which are not yet utilized within the extension) make this possible

Reply:
In this case we want to remove the 10 (or 9) last sequential empty accounts, so it should be okay if you remove them all at the end.